### PR TITLE
Fix permissions for resolv.conf and hosts

### PR DESCRIPTION
**Pull request by apostasie**
WriteFile permissions are before umask is applied. For people using aggressive umasks (0077), /etc/resolv.conf will end-up unreadable for non root processes.

Fix https://github.com/containerd/nerdctl/issues/3704

### CODE REVIEW

**Comment by apostasie** [comment]
Note: we should yank out every and all filesystem direct access away.

The Store interface did a lot of that already, and did solve atomicity and concurrency, but this is yet another proof that letting people use the filesystem directly is almost always a mistake and we should provide safer abstractions instead.

File: pkg/dnsutil/hostsstore/hostsstore.go [hunk 1]
```diff
@@ -114,6 +114,9 @@ func (x *hostsStore) Acquire(meta Meta) (err error) {
 		if err = os.WriteFile(loc, []byte{}, 0o644); err != nil {
 			return errors.Join(store.ErrSystemFailure, err)
 		}
+		if err = os.Chmod(loc, 0o644); err != nil {
```
Comment (AkihiroSuda) [inline pkg/dnsutil/hostsstore/hostsstore.go:117]: Could you add a comment to explain why `0o644` in the arg of `WriteFile` doesn't take an expected effect?

File: pkg/dnsutil/hostsstore/hostsstore.go [hunk 2]
```diff
@@ -175,6 +178,9 @@ func (x *hostsStore) AllocHostsFile(id string, content []byte) (location string,
 		if err != nil {
 			err = errors.Join(store.ErrSystemFailure, err)
 		}
+		if err = os.Chmod(loc, 0o644); err != nil {
```
Comment (AkihiroSuda) [inline pkg/dnsutil/hostsstore/hostsstore.go:181]: same as above

File: pkg/resolvconf/resolvconf.go [hunk 3]
```diff
@@ -317,7 +317,12 @@ func Build(path string, dns, dnsSearch, dnsOptions []string) (*File, error) {
 		return nil, err
 	}
 
-	return &File{Content: content.Bytes(), Hash: hash}, os.WriteFile(path, content.Bytes(), 0644)
+	err = os.WriteFile(path, content.Bytes(), 0o644)
+	if err != nil {
+		return nil, err
+	}
+
+	return &File{Content: content.Bytes(), Hash: hash}, os.Chmod(path, 0o644)
```
Comment (AkihiroSuda) [inline pkg/resolvconf/resolvconf.go:325]: same as above

File: pkg/dnsutil/hostsstore/hostsstore.go (hunk 1 above)
Comment (apostasie) [reply pkg/dnsutil/hostsstore/hostsstore.go:118]: Done

File: pkg/dnsutil/hostsstore/hostsstore.go [hunk 4]
```diff
@@ -114,6 +114,10 @@ func (x *hostsStore) Acquire(meta Meta) (err error) {
 		if err = os.WriteFile(loc, []byte{}, 0o644); err != nil {
 			return errors.Join(store.ErrSystemFailure, err)
 		}
+		// See https://www.man7.org/linux/man-pages/man2/open.2.html
```
Comment (AkihiroSuda) [inline pkg/dnsutil/hostsstore/hostsstore.go:117]: This doesn't explain how the umask rule differs across Go's `os.WriteFile` and `os.Chmod`

Comment (apostasie) [reply pkg/dnsutil/hostsstore/hostsstore.go]: No chmod implementation (system or language) would ever honor change the provided mode to account for umask - that would not make sense.

As for WriteFile, it calls `syscall.Open` underneath - and Open behavior wrt umask is covered in the man page.

Anyhow, changed the comment.

**Review by AkihiroSuda — APPROVED** [review]
Thanks

**Pull request merged**